include/os/linux/kernel/linux/simd_x86.h

```diff
@@ -126,38 +126,64 @@
 #endif
 
 #else /* defined(KERNEL_EXPORTS_X86_FPU) */
+
 /*
  * When the kernel_fpu_* symbols are unavailable then provide our own
  * versions which allow the FPU to be safely used in kernel threads.
  * In practice, this is not a significant restriction for ZFS since the
  * vast majority of SIMD operations are performed by the IO pipeline.
  */
+#if defined(HAVE_KERNEL_FPU_INTERNAL)
 
 /*
- * Returns non-zero if FPU operations are allowed in the current context.
+ * FPU usage only allowed in dedicated kernel threads.
  */
-#if defined(HAVE_KERNEL_TIF_NEED_FPU_LOAD)
-#define	kfpu_allowed()		((current->flags & PF_KTHREAD) && \
-				test_thread_flag(TIF_NEED_FPU_LOAD))
-#elif defined(HAVE_KERNEL_FPU_INITIALIZED)
-#define	kfpu_allowed()		((current->flags & PF_KTHREAD) && \
-				current->thread.fpu.initialized)
-#else
-#define	kfpu_allowed()		0
-#endif
+#define	kfpu_allowed()		(current->flags & PF_KTHREAD)
+#define	ex_handler_fprestore	ex_handler_default
```

Comment:
this single define might be enough to just allow use to call
copy_fpregs_to_kernel in kfpu_begin
copy_kernel_to_fpregs in kfpu_end (with missing exception handling, but that is missing from the current code in master and from this PR anyway ;))

it might be an easier-to-understand quick fix for master?

Reply:
Unfortunately it's not since the expansion of the `kernel_insn` macro occurs in the static inline function.  That's probably for the best since doing the save/restore ourselves ensures any future kernel changes in this area won't effect this code.